File: crates/brainwires-stores/src/memory/message_store.rs
use anyhow::{Context, Result};
use std::sync::Arc;

use brainwires_storage::CachedEmbeddingProvider;
use brainwires_storage::databases::{
    FieldDef, FieldType, FieldValue, Filter, Record, StorageBackend, record_get,
};

const TABLE_NAME: &str = "messages";

/// Metadata for a message
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct MessageMetadata {
    /// Unique message identifier.
    pub message_id: String,
    /// Conversation this message belongs to.
    pub conversation_id: String,
    /// Message role (e.g., "user", "assistant").
    pub role: String,
    /// Message text content.
    pub content: String,
    /// Token count estimate.
    pub token_count: Option<i32>,
    /// Model that generated this message.
    pub model_id: Option<String>,
    /// Image references as JSON array string.
    pub images: Option<String>, // JSON array as string
    /// Creation timestamp (Unix seconds).
    pub created_at: i64,
    /// Optional Unix timestamp after which this entry should be evicted.
    ///
    /// `None` means no expiry (the entry persists indefinitely).  Use
    /// [`MessageStore::delete_expired`] to perform bulk eviction, or call
    /// `TieredMemory::evict_expired` (in `brainwires-memory`) for tier-aware cleanup.
    pub expires_at: Option<i64>,
}

/// Return the backend-agnostic table schema for messages.
fn table_schema(embedding_dim: usize) -> Vec<FieldDef> {
    vec![
        FieldDef::required("vector", FieldType::Vector(embedding_dim)),
        FieldDef::required("message_id", FieldType::Utf8),
        FieldDef::required("conversation_id", FieldType::Utf8),
        FieldDef::required("role", FieldType::Utf8),
        FieldDef::required("content", FieldType::Utf8),
        FieldDef::optional("token_count", FieldType::Int32),
        FieldDef::optional("model_id", FieldType::Utf8),
        FieldDef::optional("images", FieldType::Utf8),
        FieldDef::required("created_at", FieldType::Int64),
        FieldDef::optional("expires_at", FieldType::Int64),
    ]
}

/// Arrow `Schema` for the messages table (LanceDatabase compatibility).
pub fn messages_schema(embedding_dim: usize) -> Arc<arrow_schema::Schema> {
    use arrow_schema::{DataType, Field, Schema};

    Arc::new(Schema::new(vec![
        Field::new(
            "vector",
            DataType::FixedSizeList(
                Arc::new(Field::new("item", DataType::Float32, true)),
                embedding_dim as i32,
            ),
            false,
        ),
        Field::new("message_id", DataType::Utf8, false),
        Field::new("conversation_id", DataType::Utf8, false),
        Field::new("role", DataType::Utf8, false),
        Field::new("content", DataType::Utf8, false),
        Field::new("token_count", DataType::Int32, true),
        Field::new("model_id", DataType::Utf8, true),
        Field::new("images", DataType::Utf8, true),
        Field::new("created_at", DataType::Int64, false),
        Field::new("expires_at", DataType::Int64, true),
    ]))
}

fn to_record(m: &MessageMetadata, embedding: Vec<f32>) -> Record {
    vec![
        ("vector".into(), FieldValue::Vector(embedding)),
        (
            "message_id".into(),
            FieldValue::Utf8(Some(m.message_id.clone())),
        ),
        (
            "conversation_id".into(),
            FieldValue::Utf8(Some(m.conversation_id.clone())),
        ),
        ("role".into(), FieldValue::Utf8(Some(m.role.clone()))),
        ("content".into(), FieldValue::Utf8(Some(m.content.clone()))),
        ("token_count".into(), FieldValue::Int32(m.token_count)),
        ("model_id".into(), FieldValue::Utf8(m.model_id.clone())),
        ("images".into(), FieldValue::Utf8(m.images.clone())),
        ("created_at".into(), FieldValue::Int64(Some(m.created_at))),
        ("expires_at".into(), FieldValue::Int64(m.expires_at)),
    ]
}

fn from_record(r: &Record) -> Result<MessageMetadata> {
    Ok(MessageMetadata {
        message_id: record_get(r, "message_id")
            .and_then(|v| v.as_str())
            .context("missing message_id")?
            .to_string(),
        conversation_id: record_get(r, "conversation_id")
            .and_then(|v| v.as_str())
            .context("missing conversation_id")?
            .to_string(),
        role: record_get(r, "role")
            .and_then(|v| v.as_str())
            .context("missing role")?
            .to_string(),
        content: record_get(r, "content")
            .and_then(|v| v.as_str())
            .context("missing content")?
            .to_string(),
        token_count: record_get(r, "token_count").and_then(|v| v.as_i32()),
        model_id: record_get(r, "model_id")
            .and_then(|v| v.as_str())
            .map(String::from),
        images: record_get(r, "images")
            .and_then(|v| v.as_str())
            .map(String::from),
        created_at: record_get(r, "created_at")
            .and_then(|v| v.as_i64())
            .context("missing created_at")?,
        expires_at: record_get(r, "expires_at").and_then(|v| v.as_i64()),
    })
}

/// Store for managing messages with semantic search
pub struct MessageStore<B: StorageBackend = brainwires_storage::databases::lance::LanceDatabase> {
    backend: Arc<B>,
    embeddings: Arc<CachedEmbeddingProvider>,
}

impl<B: StorageBackend> MessageStore<B> {
    /// Create a new message store
    pub fn new(backend: Arc<B>, embeddings: Arc<CachedEmbeddingProvider>) -> Self {
        Self {
            backend,
            embeddings,
        }
    }

    /// Ensure the underlying table exists.
    pub async fn ensure_table(&self) -> Result<()> {
        self.backend
            .ensure_table(TABLE_NAME, &table_schema(self.embeddings.dimension()))
            .await
    }

    /// Add a message to the store
    pub async fn add(&self, message: MessageMetadata) -> Result<()> {
        // Generate embedding for the content
        let embedding = self.embeddings.embed(&message.content)?;
        let record = to_record(&message, embedding);

        self.backend
            .insert(TABLE_NAME, vec![record])
            .await
            .context("Failed to add message")?;

        Ok(())
    }

    /// Add multiple messages in batch
    pub async fn add_batch(&self, messages: Vec<MessageMetadata>) -> Result<()> {
        if messages.is_empty() {
            return Ok(());
        }

        // Generate embeddings for all messages
        let contents: Vec<String> = messages.iter().map(|m| m.content.clone()).collect();
        let embeddings = self.embeddings.embed_batch(&contents)?;

        let records: Vec<Record> = messages
            .iter()
            .zip(embeddings.into_iter())
            .map(|(m, emb)| to_record(m, emb))
            .collect();

        self.backend
            .insert(TABLE_NAME, records)
            .await
            .context("Failed to add messages")?;

        Ok(())
    }

    /// Get a single message by ID
    pub async fn get(&self, message_id: &str) -> Result<Option<MessageMetadata>> {
        let filter = Filter::Eq(
            "message_id".into(),
            FieldValue::Utf8(Some(message_id.to_string())),
        );
        let records = self
            .backend
            .query(TABLE_NAME, Some(&filter), Some(1))
            .await?;

        match records.first() {
            Some(r) => Ok(Some(from_record(r)?)),
            None => Ok(None),
        }
    }

    /// Get messages for a conversation
    pub async fn get_by_conversation(&self, conversation_id: &str) -> Result<Vec<MessageMetadata>> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        let records = self.backend.query(TABLE_NAME, Some(&filter), None).await?;

        records.iter().map(from_record).collect()
    }

    /// Search messages by semantic similarity
    pub async fn search(
        &self,
        query: &str,
        limit: usize,
        min_score: f32,
    ) -> Result<Vec<(MessageMetadata, f32)>> {
        self.search_with_filter(query, limit, min_score, None).await
    }

    /// Search messages within a specific conversation by semantic similarity
    pub async fn search_conversation(
        &self,
        conversation_id: &str,
        query: &str,
        limit: usize,
        min_score: f32,
    ) -> Result<Vec<(MessageMetadata, f32)>> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        self.search_with_filter(query, limit, min_score, Some(filter))
            .await
    }

    /// Search messages with optional filter by semantic similarity
    async fn search_with_filter(
        &self,
        query: &str,
        limit: usize,
        min_score: f32,
        filter: Option<Filter>,
    ) -> Result<Vec<(MessageMetadata, f32)>> {
        // Generate query embedding (use cached version for repeated queries)
        let query_embedding = self.embeddings.embed_cached(query)?;

        let scored = self
            .backend
            .vector_search(
                TABLE_NAME,
                "vector",
                query_embedding,
                limit,
                filter.as_ref(),
            )
            .await?;

        let mut messages_with_scores = Vec::new();

        for sr in scored {
            if sr.score >= min_score {
                let message = from_record(&sr.record)?;
                messages_with_scores.push((message, sr.score));
            }
        }

        Ok(messages_with_scores)
    }

    /// Delete all messages for a conversation
    pub async fn delete_by_conversation(&self, conversation_id: &str) -> Result<()> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        self.backend.delete(TABLE_NAME, &filter).await?;
        Ok(())
    }

    /// Delete a specific message
    pub async fn delete(&self, message_id: &str) -> Result<()> {
        let filter = Filter::Eq(
            "message_id".into(),
            FieldValue::Utf8(Some(message_id.to_string())),
        );
        self.backend.delete(TABLE_NAME, &filter).await?;
        Ok(())
    }

    /// Delete all messages whose `expires_at` timestamp is in the past.
    ///
    /// Returns the number of rows deleted.  Rows with `expires_at = NULL`
    /// (no TTL) are never touched.
    ///
    /// Call this at agent run completion or on a periodic background schedule
    /// to enforce session-tier TTL policies.
    pub async fn delete_expired(&self) -> Result<usize> {
        use chrono::Utc;
        let now = Utc::now().timestamp();

        let filter = Filter::And(vec![
            Filter::NotNull("expires_at".into()),
            Filter::Lte("expires_at".into(), FieldValue::Int64(Some(now))),
        ]);

        let count = self.backend.count(TABLE_NAME, Some(&filter)).await?;
        if count > 0 {
            self.backend.delete(TABLE_NAME, &filter).await?;
        }
        Ok(count)
    }
}

File: deno/packages/stores/message_store.ts
/**
 * Message Store -- persists conversation messages with semantic search.
 *
 * Equivalent to Rust's `stores/message_store.rs` in rullama-storage.
 * @module
 */

import type { EmbeddingProvider } from "@rullama/core";
import type { StorageBackend } from "@rullama/storage";
import {
  type FieldDef,
  FieldTypes,
  fieldValueAsI32,
  fieldValueAsI64,
  fieldValueAsStr,
  FieldValues,
  type Filter,
  Filters,
  optionalField,
  type Record,
  recordGet,
  requiredField,
} from "@rullama/storage";

const TABLE_NAME = "messages";

/** Metadata for a message. */
export interface MessageMetadata {
  /** Unique message identifier; the key `get` and `delete` use. */
  messageId: string;
  /** Id of the conversation the message belongs to. */
  conversationId: string;
  /** Author role of the message (e.g. `"user"`, `"assistant"`, `"system"`). */
  role: string;
  /** Message text; this is what gets embedded for semantic search. */
  content: string;
  /** Token count of `content`, if the caller measured it. */
  tokenCount?: number;
  /** Identifier of the model that produced the message, if any. */
  modelId?: string;
  /** Image attachments serialized by the caller into one string (stored opaquely), if any. */
  images?: string;
  /** Creation time as a Unix timestamp in seconds. */
  createdAt: number;
  /** Expiry time as a Unix timestamp in seconds; once reached, `deleteExpired` removes the message. */
  expiresAt?: number;
}

function tableSchema(embeddingDim: number): FieldDef[] {
  return [
    requiredField("vector", FieldTypes.Vector(embeddingDim)),
    requiredField("message_id", FieldTypes.Utf8),
    requiredField("conversation_id", FieldTypes.Utf8),
    requiredField("role", FieldTypes.Utf8),
    requiredField("content", FieldTypes.Utf8),
    optionalField("token_count", FieldTypes.Int32),
    optionalField("model_id", FieldTypes.Utf8),
    optionalField("images", FieldTypes.Utf8),
    requiredField("created_at", FieldTypes.Int64),
    optionalField("expires_at", FieldTypes.Int64),
  ];
}

function toRecord(m: MessageMetadata, embedding: number[]): Record {
  return [
    ["vector", FieldValues.Vector(embedding)],
    ["message_id", FieldValues.Utf8(m.messageId)],
    ["conversation_id", FieldValues.Utf8(m.conversationId)],
    ["role", FieldValues.Utf8(m.role)],
    ["content", FieldValues.Utf8(m.content)],
    ["token_count", FieldValues.Int32(m.tokenCount ?? null)],
    ["model_id", FieldValues.Utf8(m.modelId ?? null)],
    ["images", FieldValues.Utf8(m.images ?? null)],
    ["created_at", FieldValues.Int64(m.createdAt)],
    ["expires_at", FieldValues.Int64(m.expiresAt ?? null)],
  ];
}

function fromRecord(r: Record): MessageMetadata {
  const messageId = recordGet(r, "message_id");
  const conversationId = recordGet(r, "conversation_id");
  const role = recordGet(r, "role");
  const content = recordGet(r, "content");
  const createdAt = recordGet(r, "created_at");

  if (!messageId || !conversationId || !role || !content || !createdAt) {
    throw new Error("Missing required fields in message record");
  }

  return {
    messageId: fieldValueAsStr(messageId)!,
    conversationId: fieldValueAsStr(conversationId)!,
    role: fieldValueAsStr(role)!,
    content: fieldValueAsStr(content)!,
    tokenCount: recordGet(r, "token_count")
      ? fieldValueAsI32(recordGet(r, "token_count")!)
      : undefined,
    modelId: recordGet(r, "model_id")
      ? fieldValueAsStr(recordGet(r, "model_id")!)
      : undefined,
    images: recordGet(r, "images")
      ? fieldValueAsStr(recordGet(r, "images")!)
      : undefined,
    createdAt: fieldValueAsI64(createdAt)!,
    expiresAt: recordGet(r, "expires_at")
      ? fieldValueAsI64(recordGet(r, "expires_at")!)
      : undefined,
  };
}

/** Interface for message store operations. */
export interface MessageStoreI {
  /** Create the backing `messages` table (including its embedding vector column) if it does not already exist. */
  ensureTable(): Promise<void>;
  /** Persist one message, embedding its `content` for semantic search. */
  add(message: MessageMetadata): Promise<void>;
  /** Persist several messages in one insert, embedding their contents in a single batch; a no-op for an empty array. */
  addBatch(messages: MessageMetadata[]): Promise<void>;
  /** Look up a message by id; `undefined` when none has that id. */
  get(messageId: string): Promise<MessageMetadata | undefined>;
  /** Every message in a conversation, in the order the backend returns them (no sorting applied). */
  getByConversation(conversationId: string): Promise<MessageMetadata[]>;
  /**
   * Semantic search over all messages.
   * @param query Text that is embedded and compared against stored message vectors.
   * @param limit Maximum number of hits requested from the backend.
   * @param minScore Hits scoring below this similarity are dropped.
   * @returns `[message, similarityScore]` pairs in the backend's ranking order.
   */
  search(
    query: string,
    limit: number,
    minScore: number,
  ): Promise<[MessageMetadata, number][]>;
  /**
   * Semantic search restricted to one conversation.
   * @param conversationId Only messages of this conversation are candidates.
   * @param query Text that is embedded and compared against stored message vectors.
   * @param limit Maximum number of hits requested from the backend.
   * @param minScore Hits scoring below this similarity are dropped.
   * @returns `[message, similarityScore]` pairs in the backend's ranking order.
   */
  searchConversation(
    conversationId: string,
    query: string,
    limit: number,
    minScore: number,
  ): Promise<[MessageMetadata, number][]>;
  /** Delete every message belonging to a conversation. */
  deleteByConversation(conversationId: string): Promise<void>;
  /** Delete one message by id (no-op when absent). */
  delete(messageId: string): Promise<void>;
  /**
   * Delete every message whose `expiresAt` is set and not later than now.
   * @returns The number of messages removed.
   */
  deleteExpired(): Promise<number>;
}

/** Store for managing messages with semantic search. */
export class MessageStore implements MessageStoreI {
  /**
   * Create a store over `backend`; call {@link ensureTable} before use.
   * @param backend Storage backend that holds the `messages` table.
   * @param embeddings Provider used to embed message content and search
   *   queries; its `dimension` fixes the width of the table's vector column.
   */
  constructor(
    private readonly backend: StorageBackend,
    private readonly embeddings: EmbeddingProvider,
  ) {}

  /** Create the `messages` table on the backend, sized to the embedding provider's `dimension`. */
  async ensureTable(): Promise<void> {
    await this.backend.ensureTable(
      TABLE_NAME,
      tableSchema(this.embeddings.dimension),
    );
  }

  /** Embed `message.content` and insert the message as one record. */
  async add(message: MessageMetadata): Promise<void> {
    const embedding = await this.embeddings.embed(message.content);
    const record = toRecord(message, embedding);
    await this.backend.insert(TABLE_NAME, [record]);
  }

  /** Embed all contents with one `embedBatch` call and insert the records together; returns early for an empty array. */
  async addBatch(messages: MessageMetadata[]): Promise<void> {
    if (messages.length === 0) return;
    const contents = messages.map((m) => m.content);
    const embeddings = await this.embeddings.embedBatch(contents);
    const records = messages.map((m, i) => toRecord(m, embeddings[i]));
    await this.backend.insert(TABLE_NAME, records);
  }

  /** Query the backend for the single record whose `message_id` matches. */
  async get(messageId: string): Promise<MessageMetadata | undefined> {
    const filter = Filters.Eq("message_id", FieldValues.Utf8(messageId));
    const records = await this.backend.query(TABLE_NAME, filter, 1);
    return records.length > 0 ? fromRecord(records[0]) : undefined;
  }

  /** Query every record whose `conversation_id` matches, in backend order. */
  async getByConversation(conversationId: string): Promise<MessageMetadata[]> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    const records = await this.backend.query(TABLE_NAME, filter);
    return records.map(fromRecord);
  }

  /** Vector search over the whole table; see {@link searchWithFilter}. */
  // deno-lint-ignore require-await
  async search(
    query: string,
    limit: number,
    minScore: number,
  ): Promise<[MessageMetadata, number][]> {
    return this.searchWithFilter(query, limit, minScore);
  }

  /** Vector search filtered to one `conversation_id`; see {@link searchWithFilter}. */
  // deno-lint-ignore require-await
  async searchConversation(
    conversationId: string,
    query: string,
    limit: number,
    minScore: number,
  ): Promise<[MessageMetadata, number][]> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    return this.searchWithFilter(query, limit, minScore, filter);
  }

  /**
   * Embed `query`, run the backend's vector search on the `vector` column
   * (optionally narrowed by `filter`), and drop hits scoring below `minScore`.
   */
  private async searchWithFilter(
    query: string,
    limit: number,
    minScore: number,
    filter?: Filter,
  ): Promise<[MessageMetadata, number][]> {
    const queryEmbedding = await this.embeddings.embed(query);
    const scored = await this.backend.vectorSearch(
      TABLE_NAME,
      "vector",
      queryEmbedding,
      limit,
      filter,
    );

    return scored
      .filter((sr) => sr.score >= minScore)
      .map((sr) => [fromRecord(sr.record), sr.score]);
  }

  /** Delete every record whose `conversation_id` matches. */
  async deleteByConversation(conversationId: string): Promise<void> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    await this.backend.delete(TABLE_NAME, filter);
  }

  /** Delete the record whose `message_id` matches. */
  async delete(messageId: string): Promise<void> {
    const filter = Filters.Eq("message_id", FieldValues.Utf8(messageId));
    await this.backend.delete(TABLE_NAME, filter);
  }

  /**
   * Count records with a non-null `expires_at` <= now, delete them if there
   * are any, and return that count.
   */
  async deleteExpired(): Promise<number> {
    const now = Math.floor(Date.now() / 1000);
    const filter = Filters.And([
      Filters.NotNull("expires_at"),
      Filters.Lte("expires_at", FieldValues.Int64(now)),
    ]);

    const count = await this.backend.count(TABLE_NAME, filter);
    if (count > 0) {
      await this.backend.delete(TABLE_NAME, filter);
    }
    return count;
  }
}

/**
 * In-memory MessageStore for testing (does not require an embedding provider
 * -- stores a zero vector).
 */
export class InMemoryMessageStore implements MessageStoreI {
  private messages: Map<string, MessageMetadata> = new Map();

  /** No-op: there is no table to create in memory. */
  async ensureTable(): Promise<void> {
    await Promise.resolve();
  }

  /** Store a shallow copy of the message keyed by `messageId`, replacing any existing one. */
  async add(message: MessageMetadata): Promise<void> {
    this.messages.set(message.messageId, { ...message });
    await Promise.resolve();
  }

  /** Store a shallow copy of each message keyed by `messageId`. */
  async addBatch(messages: MessageMetadata[]): Promise<void> {
    for (const m of messages) {
      this.messages.set(m.messageId, { ...m });
    }
    await Promise.resolve();
  }

  /** Return the stored message for `messageId`, if any. */
  async get(messageId: string): Promise<MessageMetadata | undefined> {
    return await Promise.resolve(this.messages.get(messageId));
  }

  /** Stored messages whose `conversationId` matches, in insertion order. */
  async getByConversation(conversationId: string): Promise<MessageMetadata[]> {
    return await Promise.resolve(
      [...this.messages.values()].filter((m) =>
        m.conversationId === conversationId
      ),
    );
  }

  /**
   * Returns the first `limit` stored messages, each with a score of 1.0; the
   * query and `minScore` are ignored because nothing is embedded in memory.
   */
  async search(
    _query: string,
    limit: number,
    _minScore: number,
  ): Promise<[MessageMetadata, number][]> {
    // Simple: return first N messages with score 1.0
    return await Promise.resolve(
      [...this.messages.values()].slice(0, limit).map((m) =>
        [m, 1.0] as [MessageMetadata, number]
      ),
    );
  }

  /**
   * Returns the first `limit` messages of the conversation, each with a score
   * of 1.0; the query and `minScore` are ignored.
   */
  async searchConversation(
    conversationId: string,
    _query: string,
    limit: number,
    _minScore: number,
  ): Promise<[MessageMetadata, number][]> {
    return await Promise.resolve(
      [...this.messages.values()]
        .filter((m) => m.conversationId === conversationId)
        .slice(0, limit)
        .map((m) => [m, 1.0] as [MessageMetadata, number]),
    );
  }

  /** Remove every stored message whose `conversationId` matches. */
  async deleteByConversation(conversationId: string): Promise<void> {
    for (const [id, m] of this.messages) {
      if (m.conversationId === conversationId) this.messages.delete(id);
    }
    await Promise.resolve();
  }

  /** Remove the message from the map (no-op when absent). */
  async delete(messageId: string): Promise<void> {
    this.messages.delete(messageId);
    await Promise.resolve();
  }

  /** Remove every message whose `expiresAt` is set and <= now; returns how many were removed. */
  async deleteExpired(): Promise<number> {
    const now = Math.floor(Date.now() / 1000);
    let count = 0;
    for (const [id, m] of this.messages) {
      if (m.expiresAt !== undefined && m.expiresAt <= now) {
        this.messages.delete(id);
        count++;
      }
    }
    return await Promise.resolve(count);
  }
}
